cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,823 @@
+        b: BindingId,
+        /// If the test succeeds, evaluate this block.
+        body: Block,
+    },
+    /// Evaluate a block once with each value of the given binding site.
+    Loop {
+        /// A binding site of type [Binding::Iterator]. Its source binding site
+        /// must be a multi-extractor or multi-constructor call.
+        result: BindingId,
+        /// What to evaluate with each binding.
+        body: Block,
+    },
+    /// Return a result from the right-hand side of a rule. If we're building a
+    /// multi-constructor then this doesn't actually return, but adds to a list
+    /// of results instead. Otherwise this return stops evaluation before any
+    /// later steps.
+    Return {
+        /// Where was the rule defined that had this right-hand side?
+        pos: Pos,
+        /// What is the result expression which should be returned if this
+        /// rule matched?
+        result: BindingId,
+    },
+}
+
+/// One concrete pattern and the block to evaluate if the pattern matches.
+pub struct MatchArm {
+    /// The pattern to match.
+    pub constraint: Constraint,
+    /// If this pattern matches, it brings these bindings into scope. If a
+    /// binding is unused in this block, then the corresponding position in the
+    /// pattern's bindings may be `None`.
+    pub bindings: Vec<Option<BindingId>>,
+    /// Steps to evaluate if the pattern matched.
+    pub body: Block,
+}
+
+/// Given a set of rules that's been partitioned into two groups, move rules
+/// from the first partition to the second if there are higher-priority rules
+/// in the second group. In the final generated code, we'll check the rules
+/// in the first ("selected") group before any in the second ("deferred")
+/// group. But we need the result to be _as if_ we checked the rules in strict
+/// descending priority order.
+///
+/// When evaluating the relationship between one rule in the selected set and
+/// one rule in the deferred set, there are two cases where we can keep a rule
+/// in the selected set:
+/// - The deferred rule is lower priority than the selected rule; or
+/// - The two rules don't overlap, meaning they can't match on the same inputs.
+/// In either case, if the selected rule matches then we know the deferred rule
+/// would not have been the one we wanted anyway; and if it doesn't match then
+/// the fall-through semantics of the code we generate will let us go on to
+/// check the deferred rule.
+///
+/// So a rule can stay in the selected set as long as it's in one of the above
+/// relationships with every rule in the deferred set.
+///
+/// In practice only the priority matters. Checking overlap here doesn't change
+/// the output on any backend as of this writing. But I've measured and it
+/// hardly takes any time, so I'm leaving it in just in case somebody writes
+/// rules someday where it helps.
+fn respect_priority(rules: &RuleSet, order: &mut [usize], partition_point: usize) -> usize {
```

Comment:
Is it worth stating here that the returned index will always be `<= partition_point`?